app/src/androidTest/java/org/mozilla/focus/activity/ContextMenusTest.kt

```diff
@@ -36,41 +40,130 @@ class ContextMenusTest {
+
+        searchScreen {
+        }.loadPage(imagesTestPage.url) {
+            longPressLink("download icon")
+            verifyImageContextMenu(true, imageAssetUrl)
+        }
+    }
+
+    @SmokeTest
+    @Test
+    fun simpleImageContextMenuItemsTest() {
+        val imagesTestPage = getImageTestAsset(webServer)
+        val imageAssetUrl = webServer.url("rabbit.jpg").toString()
+
+        searchScreen {
+        }.loadPage(imagesTestPage.url) {
+            longPressLink("rabbit.jpg")
+            verifyImageContextMenu(false, imageAssetUrl)
+        }
+    }
+
+    @SmokeTest
+    @Test
+    fun linkContextMenuItemsTest() {
+        val tab1Page = getGenericTabAsset(webServer, 1)
+        val tab2Page = getGenericTabAsset(webServer, 2)
+
+        searchScreen {
+        }.loadPage(tab1Page.url) {
+            verifyPageContent("Tab 1")
+            longPressLink("Tab 2")
+            verifyLinkContextMenu(tab2Page.url)
+        }
     }
 
     @SmokeTest
     @Test
     fun copyLinkAddressTest() {
-        val tab1Url = webServer.url("tab1.html").toString()
-        val tab2Url = webServer.url("tab2.html").toString()
+        val tab1Page = getGenericTabAsset(webServer, 1)
+        val tab2Page = getGenericTabAsset(webServer, 2)
 
         searchScreen {
-        }.loadPage(tab1Url) {
+        }.loadPage(tab1Page.url) {
             verifyPageContent("Tab 1")
             longPressLink("Tab 2")
-            verifyLinkContextMenu(tab2Url)
+            verifyLinkContextMenu(tab2Page.url)
             clickContextMenuCopyLink()
         }
 
         searchScreen {
             clickToolbar()
             clearSearchBar()
             longPressSearchBar()
-            pasteAndLoadLink()
+        }.pasteAndLoadLink {
+            progressBar.waitUntilGone(TestHelper.waitingTime)
+            verifyPageURL(tab2Page.url)
         }
+    }
+
+    @Test
+    fun copyImageLocationTest() {
+        val imagesTestPage = getImageTestAsset(webServer)
+        val imageAssetUrl = webServer.url("rabbit.jpg").toString()
 
-        browserScreen {
+        searchScreen {
+        }.loadPage(imagesTestPage.url) {
+            longPressLink("rabbit.jpg")
+            verifyImageContextMenu(false, imageAssetUrl)
+            clickCopyImageLocation()
+        }.openSearchBar {
+            clearSearchBar()
+            longPressSearchBar()
+        }.pasteAndLoadLink {
             progressBar.waitUntilGone(TestHelper.waitingTime)
-            verifyPageURL(tab2Url)
+            verifyPageURL(imageAssetUrl)
+        }
+    }
+
+    @SmokeTest
+    @Test
+    fun saveImageTest() {
+        val imagesTestPage = getImageTestAsset(webServer)
+        fileName = "rabbit.jpg"
+
+        searchScreen {
+        }.loadPage(imagesTestPage.url) {
+            longPressLink(fileName)
+        }.clickSaveImage {
+            // If permission dialog appears, grant it
+            if (permAllowBtn.exists()) {
+                permAllowBtn.click()
+            }
```

Comment:
I think it should be safe to remove this code on API 30 and use the same runtime grant permissions, worth checking out
``     @Rule
     public GrantPermissionRule permissionRule = GrantPermissionRule.grant(Manifest.permission.WRITE_EXTERNAL_STORAGE);
``
https://developer.android.com/reference/android/Manifest.permission.html#READ_EXTERNAL_STORAGE/

Reply:
done